File: lib/afterglow/api_actions/api_action.ex
defmodule AfterGlow.ApiActions.ApiAction do
  use Ecto.Schema
  import Ecto.Changeset
  alias AfterGlow.Question
  alias AfterGlow.Repo
  import EctoEnum, only: [defenum: 2]

  defenum(MethodEnum, GET: 1, POST: 2, PUT: 3, DELETE: 4, PATCH: 5)
  defenum(ActionLevelEnum, question_response: 1, question: 2)

  @derive {Jason.Encoder,
           only: [
             :id,
             :question_id,
             :loading_message,
             :url,
             :headers,
             :body,
             :method,
             :name,
             :column,
             :color,
             :hidden,
             :open_in_new_tab,
             :response_settings,
             :visualization_id,
             :display_settings,
      :open_option
           ]}
  schema "api_actions" do
    belongs_to(:question, Question)
    belongs_to(:visualization, Question)
    field(:url, :string)
    field(:headers, :map)
    field(:body, :string)
    field(:method, MethodEnum)
    field(:action_level, ActionLevelEnum)
    field(:name, :string)
    field(:column, :string)
    field(:color, :string)
    field(:loading_message, :string)
    field(:hidden, :boolean)
    field(:open_in_new_tab, :boolean)
    field(:response_settings, :map)
    field(:display_settings, :map)
    field(:open_option, :string)
    timestamps()
  end

  def changeset(%__MODULE__{} = api_actions, attrs) do
    api_actions
    |> cast(attrs, [
      :url,
      :name,
      :headers,
      :body,
      :color,
      :open_in_new_tab,
      :method,
      :question_id,
      :visualization_id,
      :action_level,
      :loading_message,
      :hidden,
      :response_settings,
      :column,
      :display_settings,
      :open_option
    ])
    |> validate_required([:url, :name, :method, :question_id])
  end

  def cache_deletable_associations do
    [:question, :visualization]
  end

  def set_default_action_level(api_action) do
    unless api_action.action_level do
      Ecto.Changeset.change(api_action,
        action_level: 1
      )
      |> Repo.update!()
    end
  end
end

File: backend/src/repository/api_action.rs
use diesel::PgConnection;

use super::models::{ActionLevel, ApiAction, ApiActionChangeset};
use super::schema::api_actions;

use diesel::prelude::*;
use diesel::result::Error;

impl ApiAction {
    pub fn to_changeset(&self) -> ApiActionChangeset {
        ApiActionChangeset {
            question_id: self.question_id,
            url: self.url.clone(),
            headers: self.headers.clone(),
            body: self.body.clone(),
            method: self.method.clone(),
            name: self.name.clone(),
            color: self.color.clone(),
            open_in_new_tab: self.open_in_new_tab.clone(),
            response_settings: self.response_settings.clone(),
            hidden: self.hidden,
            inserted_at: self.inserted_at,
            updated_at: self.updated_at,
            column: self.column.clone(),
            on_success: self.on_success,
            on_failure: self.on_failure,
            failure_message: self.failure_message.clone(),
            failure_key: self.failure_key.clone(),
            success_message: self.success_message.clone(),
            success_key: self.success_key.clone(),
            action_level: self.action_level.clone(),
            visualization_id: self.visualization_id,
            loading_message: self.loading_message.clone(),
            display_settings: self.display_settings.clone(),
            open_option: self.open_option.clone(),
        }
    }
    pub fn find_direct_actions_by_question_id(
        conn: &mut PgConnection,
        qid: i32,
    ) -> Result<Self, Error> {
        api_actions::table
            .filter(api_actions::question_id.eq(qid))
            .filter(api_actions::action_level.eq(ActionLevel::Question))
            .first::<Self>(conn)
    }
    pub fn find_by_question_id(conn: &mut PgConnection, qid: i32) -> Result<Vec<Self>, Error> {
        api_actions::table
            .filter(api_actions::question_id.eq(qid))
            .filter(api_actions::action_level.eq(ActionLevel::QuestionResponse))
            .load::<Self>(conn)
    }
}
